presto-main/src/main/java/com/facebook/presto/execution/scheduler/SectionExecutionFactory.java

```diff
@@ -335,8 +335,8 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
                     // verify execution is consistent with planner's decision on dynamic lifespan schedule
                     verify(bucketNodeMap.isDynamic() == dynamicLifespanSchedule);
 
-                    if (!bucketNodeMap.isDynamic()) {
-                        stageNodeList = ((FixedBucketNodeMap) bucketNodeMap).getBucketToNode().stream()
+                    if (!bucketNodeMap.isDynamic() || bucketNodeMap.getBucketToNode().isPresent()) {
+                        stageNodeList = bucketNodeMap.getBucketToNode().get().stream()
```

Comment:
Just out of curiosity, what would happen if `bucketNodeMap` is updated in `DynamicBucketNodeMap` (thus `stageNodeList` is not consistent with `DynamicBucketNodeMap#getBucketToNode`)?

Reply:
It would be the same behavior as we assign random stageNodeList, when we do scheduling, we would distribute bucket (the bucket that couldn't find its desired bucket) to a random node.

`bucketNodeMap` being updated in `DynamicBucketNodeMap` would only happen in `DynamicLifespanScheduler`, which means it happens when some node/task failed and affinity can't be enforced , thus we have to fall back to random choosing.

So the difference with random stageNodeList, we try best effort to put bucket to their desired node, then fall back to random if affinity can't be achieved.